notion-server/src/eth_magic.rs

```diff
@@ -0,0 +1,165 @@
+use std::default;
+use std::collections::{HashMap};
+
+#[derive(Debug)]
+pub struct EthMagicManager {
+    beacon_chain: BeaconChain,
+    shard_chains: Vec<ShardChain>,
+}
+
+impl EthMagicManager {
+    pub fn new() -> Self {
+        Self {
+            beacon_chain: BeaconChain::new(),
+            shard_chains: Vec::new(),
+        }
+    }
+
+    // Creates a new execution environment on the BeaconChain and returns the
+    // index of the created execution environment
+    pub fn add_new_execution_environment(&mut self, ee: CreateEeArgs) -> u32 {
+        let execution_environment = ExecutionEnvironment {
+            wasm_code: ee.wasm_code,
+        };
+        self.beacon_chain.execution_environments.push(execution_environment);
+        (self.beacon_chain.execution_environments.len() - 1) as u32
+    }
+
+    // Returns the index of the newly added shard chain
+    // Longer-term, can accept a config here
+    pub fn add_new_shard_chain(&mut self) -> u32 {
+        let shard_chain = ShardChain::new();
+        self.shard_chains.push(shard_chain);
+        (self.shard_chains.len() - 1) as u32
+    }
+
+    // Returns the index of the newly added shard block
+    pub fn append_shard_block(&mut self, shard_index: usize, block: CreateShardBlockArgs) -> u32 {
+        // Worth noting that in a real-world use case "sub-transactions" may be merged
+        // into one "combined" transaction before being executed / committed to a block
+//        let &mut shard_chain = &mut self.shard_chains[shard_index];
+//
+//        // Sam to implement: create the transactions and the shard block and run the transactions
+//        let shard_block = ShardBlock::new(Vec::new());
+//
+//        shard_chain.shard_blocks.push(shard_block);
+//        (shard_chain.shard_blocks.len() - 1) as u32
+
+        unimplemented!();
+
+//        let transactions = block.transactions
+//
+//        for transaction in block.transactions {
+//            // This executes everything and presumably also updates the EE State on the shard
+//            let ee = transaction.execution_environment;
+//            let input_data = transaction.data;
+//
+//            let code = self.beacon_chain.get(ee);
+//            let runtime = RootRuntime::new(&code, shard_ee_state_or_something_similar);
+//            runtime.execute(input_data);
+//        }
+
+    }
+
+    /* Getter methods still needed
+        Beacon State
+        Shard State
+        Transactions (do we want to store EE state per shard before / after each transaction?
+
+    */
+}
+
+#[derive(Debug, Default)]
+pub struct CreateEeArgs {
+    wasm_code: Vec<u8>,
+}
+
+#[derive(Debug, Default)]
+pub struct CreateShardBlockArgs {
+
+}
+
+
+#[derive(Debug, Default)]
+struct BeaconChain {
+    // There are an unbounded number of EEs that can "exist" on the beacon chain
```

Comment:
`s/are/is/`

Reply:
I actually think "are" sounds more correct.  It's similar to the statement "there are a number of problems with this approach".  While "there is a number of problems with this approach" may technically be correct (though I'm not sure it is), since "number" is technically singular, I believe almost all native English speakers would say the "there are" version of the sentence.

Obviously this is not super related to the code haha, but to me it sounds more correct as-is.